File: C++/exercise 01/SFML/Game.cpp
#include "Game.h"

Game::Game()
{
	mWindow.create(VideoMode(screenWidth, screenHeight), "SFML");
	mIsMovingDown = mIsMovingLeft = mIsMovingRight = mIsMovingUp = false;
}

Game::Game(String title, Uint32 style)
{
	mWindow.create(VideoMode(screenWidth, screenHeight), title, style);
	world = *new World();
	player = *new Player();
	player.Move(Vector2f(Game::screenWidth/2, Game::screenHeight));
}

Game::~Game()
{
	free(&world);
	free(&player);
}

void Game::run()
{
	while (mWindow.isOpen())
	{
		processEvents();
		update();
		render();
	}
}

void Game::handlePlayerInput(Keyboard::Key key, bool isPressed)
{
	switch (key)
	{
	case sf::Keyboard::A:
		mIsMovingLeft = isPressed;
		break;
	case sf::Keyboard::D:
		mIsMovingRight = isPressed;
		break;
	case sf::Keyboard::W:
		mIsMovingUp = isPressed;
		break;
	case sf::Keyboard::S:
		mIsMovingDown = isPressed;
		break;
	default:
		break;
	}
}

void Game::processEvents()
{
	Event event;

	while (mWindow.pollEvent(event))
	{
		switch (event.type)
		{
			case Event::KeyPressed:
				handlePlayerInput(event.key.code, true);
				break;
			case Event::KeyReleased:
				handlePlayerInput(event.key.code, false);
				break;
			case Event::Closed:
				mWindow.close();
				break;
			default:
				break;
		}
	}
}

void Game::update()
{
	Vector2f movement(0.f, 0.f);

	if (mIsMovingUp)
		movement.y -= 0.1f;
	if (mIsMovingDown)
		movement.y += 0.1f;
	if (mIsMovingLeft)
		movement.x -= 0.1f;
	if (mIsMovingRight)
		movement.x += 0.1f;

	if (movement != Vector2f(0.f, 0.f))
		player.Move(movement);
}

void Game::render()
{
	mWindow.clear();

	mWindow.draw(world.sky);
	mWindow.draw(world.street);
	mWindow.draw(world.forestLeft);
	mWindow.draw(world.forestRight);
	mWindow.draw(world.lineStreetLeft);
	mWindow.draw(world.lineStreetRight);

	mWindow.draw(player.body);
	mWindow.draw(player.head);
	mWindow.draw(player.legLeft);
	mWindow.draw(player.legRight);
	mWindow.draw(player.armLeft);
	mWindow.draw(player.armRight);
	mWindow.draw(player.hat);
	mWindow.draw(player.nose);

	mWindow.display();
}

File: C#/exercise 01/sfml.net/src/Game.cs
using SFML.Graphics;
using SFML.Window;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SFML.System;

namespace sfml.net.src
{
    public class Game
    {
        #region Fields

        public static readonly uint WINDOW_WIDTH = 800;
        public static readonly uint WINDOW_HEIGHT = 600;

        private RenderWindow window;
        private string windowTitle = "SFML.Net";
        private readonly uint maxFPS = 60;

        private Player player;
        private World world;

        public static Clock DeltaTime = new Clock();

        #endregion


        #region public

        public Game(string title)
        {
            this.windowTitle = title;
            this.window = new RenderWindow(new VideoMode(WINDOW_WIDTH, WINDOW_HEIGHT), windowTitle);
            this.window.SetFramerateLimit(maxFPS);

            this.player = new Player();
            this.player.Move(new Vector2f(WINDOW_WIDTH/2, WINDOW_HEIGHT/2));
            this.world = new World();

            this.window.KeyPressed += ProcessPressedEvents;
            this.window.KeyReleased += ProcessReleasedEvents;

            this.window.Closed += (sender, e) =>
            {
                Console.WriteLine("Game Over!");
                ((RenderWindow)sender).Close();
            };
        }

        public void Run()
        {
            while(window.IsOpen)
            {
                window.DispatchEvents();
                Update();
                Render();
                DeltaTime.Restart();
            }
        }

        #endregion


        #region Private


        private void ProcessPressedEvents(object sender, KeyEventArgs e)
        {
            ProcessInput(e.Code, true);
        }

        private void ProcessReleasedEvents(object sender, KeyEventArgs e)
        {
            ProcessInput(e.Code, false);
        }

        private void ProcessInput(Keyboard.Key key, bool isPressed)
        {
            if(key == Keyboard.Key.A)
            {
                this.player.IsMovingLeft = isPressed;
            }
            if(key == Keyboard.Key.S)
            {
                this.player.IsMovingDown = isPressed;
            }
            if(key == Keyboard.Key.D)
            {
                this.player.IsMovingRight = isPressed;
            }
            if(key == Keyboard.Key.W)
            {
                this.player.IsMovingUp = isPressed;
            }
        }

        private void Update()
        {
            var movement = new Vector2f();

            if (this.player.IsMovingUp)
                movement.Y -= 1f;
            if (this.player.IsMovingDown)
                movement.Y += 1f;
            if (this.player.IsMovingLeft)
                movement.X -= 1f;
            if (this.player.IsMovingRight)
                movement.X += 1f;

            this.player.Move(movement);
        }

        private void Render()
        {
            window.Clear();
            world.Display(window);
            player.Display(window);
            window.Display();
        }

        #endregion
    }
}
